062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,223 @@
+- Failed: `.spec.replicas` is updated with an invalid value, or there is a hard goal violation, or Cruise Control is not available (`generation` == `observedGeneration`).
+
+To request a replication factor change, the Topic Operator needs to authenticate and send HTTP requests to the Cruise Control's REST API.
+For that reason, we introduce the following environment variables, which will be set automatically by the Cluster Operator, or manually by the user in case of standalone deployment.
+
+| Name                                   | Type    | Default | Description                                           |
+|:---------------------------------------|---------|---------|-------------------------------------------------------|
+| STRIMZI_CRUISE_CONTROL_ENABLED         | Boolean | false   | Whether Cruise Control integration is enabled         |
+| STRIMZI_CRUISE_CONTROL_NAMESPACE       | String  | ""      | Namespace where Cruise Control is running             |
+| STRIMZI_CRUISE_CONTROL_HOSTNAME        | String  | ""      | Cruise Control hostname                               |
+| STRIMZI_CRUISE_CONTROL_PORT            | Integer | 9090    | Cruise Control port                                   |
+| STRIMZI_CRUISE_CONTROL_SSL_ENABLED     | Boolean | false   | Whether Cruise Control SSL encryption is enabled      |
+| STRIMZI_CRUISE_CONTROL_AUTH_ENABLED    | Boolean | false   | Whether Cruise Control Basic authorization is enabled |
+| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing Cruise Control truststore      |
+| STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing Cruise Control API credentials |
+| STRIMZI_CRUISE_CONTROL_RACK_ENABLED    | Boolean | false   | Whether the target Kafka cluster has rack awareness   |
+
+The topic configuration endpoint is asynchronous, which means that it validates the task, but the execution will be queued.
+Inside Cruise Control, a valid topic configuration task is first create in an `Active` state, then moved to `InExecution`, and finally to `Completed` or `CompletedWithError`.
+Only a single task can be `InExecution` state at any given time, which includes both topic configuration and rebalance tasks.
+
+When Cruise Control integration is enabled, a `ReplicationFactorManager` is created to manage the communication with the REST API.
+The `ReplicationFactorManager` stores all ongoing replication factor changes in an in-memory cache, in order to periodically check their status.
+A single replication factor change can be ongoing at any given time for a specific `KafkaTopic` resource, following replication factor changes will remain pending, and periodically retried.
+At most `max.active.user.tasks` (default: 5) replication factor changes for different `KafkaTopic` resources can be ongoing at the same time.
```

Comment:
This probably deserves some explanation why ... and also how you will handle when you have more replication factor changes then the max tasks.

Reply:
When CC queue is full, next requests will have a warning and will be periodically retried. Proposal updated.